### PR TITLE
Retry full cycle of send/receive

### PR DESCRIPTION
**Pull request by midnightercz**
When receiver fails to receive all messages and run out of retries, repeats whole cycle of send/receive

### CODE REVIEW

File: src/pubtools/sign/signers/msgsigner.py [hunk 1]
```diff
@@ -293,57 +303,75 @@ def clear_sign(self: MsgSigner, operation: ClearSignOperation) -> SigningResults
...
+                if i != self.send_retries - 1:
+                    errors.pop(0)
+                messages = _messages
+
+            elif not errors:
+                break
 
         errors = recvc._errors
         if errors:
-            signer_results.status = "error"
```
Comment (emilyzheng) [inline src/pubtools/sign/signers/msgsigner.py:336]: This change seems unnecessary.

File: src/pubtools/sign/signers/msgsigner.py [hunk 2]
```diff
@@ -384,15 +412,19 @@ def container_sign(self: MsgSigner, operation: ContainerSignOperation) -> Signin
             LOG.info(f"Using signing key alias {signing_key} for {operation.signing_key}")
 
         for digest, reference in zip(operation.digests, operation.references):
+            repo = reference.split(":")[0].split("/", 1)[1].split(":")[0]
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply src/pubtools/sign/signers/msgsigner.py:415]: first is for protocol.

[the remark this answers is not in the document]
Comment (emilyzheng) [reply src/pubtools/sign/signers/msgsigner.py:453]: Do you mean reference like this?
```
>>> reference="docker://registry/repo:tag"
>>> reference.split(":")[0].split("/", 1)[1].split(":")[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
```
And reference in signature doesn't have protocol.

File: src/pubtools/sign/signers/msgsigner.py [hunk 3]
```diff
@@ -406,55 +487,107 @@ def container_sign(self: MsgSigner, operation: ContainerSignOperation) -> Signin
...
+                for error in errors:
+                    signer_results.error_message += f"{error.name} : {error.description}\n"
+                return signing_results
+
+            # wait for receiver to finish
+            recvt.join()
+            recvt.stop()
+            received = recvc.get_received()
+
+            for x in range(self.retries):
```
Comment (emilyzheng) [inline src/pubtools/sign/signers/msgsigner.py:539]: http://10.0.150.153:32771/pub/task/2485/log/stdout.log
```
2024-04-18 07:38:04 +0000 [INFO    ] [139834387584704][1] All messages received
2024-04-18 07:38:04 +0000 [INFO    ] Disconnected, already closed
2024-04-18 07:38:04 +0000 [ERROR   ] [MsgError(name='MessagingTimeout', description='[139834387584704] Out of time when receiving messages (0/1063)', source=PN_TIMER_TASK(<proton._reactor.Task object at 0x7f2ddb12e010>)), MsgError(name='MessagingTimeout', description='[139834387584704] Out of time when receiving messages (0/1063)', source=PN_TIMER_TASK(<proton._reactor.Task object at 0x7f2ddb12ee10>))]
2024-04-18 07:38:04 +0000 [INFO    ] Retrying receiving 0/20
```
Retrying after `All messages received`?

File: src/pubtools/sign/signers/msgsigner.py [hunk 4]
```diff
@@ -406,55 +487,107 @@ def container_sign(self: MsgSigner, operation: ContainerSignOperation) -> Signin
...
+                errors = recvc.get_errors()
+                LOG.error(errors)
```
Comment (emilyzheng) [inline src/pubtools/sign/signers/msgsigner.py:541]: Remove this?

**Review by emilyzheng — APPROVED** [review]

**Pull request merged**